doc/types.go

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package doc
+
+// ID represents a document's natural identifier. It's treated as immutable
+// by the indexing subsystem.
+type ID []byte
+
+// FieldValueType represents the different types of field values supported.
+type FieldValueType byte
+
+const (
+	// UnknownValueType demarks an unknown value type.
+	UnknownValueType FieldValueType = iota
+
+	// StringValueType demarks string value types.
+	StringValueType
+)
+
+// Term represents the value of a field.
+type Term []byte
+
+// Field represents a document field.
+// NB(prateek): The value of a Field is a single Term, this deviates
+// from conventional use of `Term` in other indexing software (e.g. Lucene).
+type Field struct {
+	Name      []byte
+	Value     Term
+	ValueType FieldValueType
+}
+
+// Document represents a document to be indexed.
+type Document struct {
+	// ID is the natural identifier for the document.
+	ID ID
+
+	// IndexableFields contains the list of fields by which to index the document.
+	IndexableFields []Field
```

Comment:
Was looking through some Lucene documentation some more and they actually distinguish three properties of fields: indexed, analyzed (we don't need to worry about this for now, though it can't hurt to think about) and stored. Indexed determines only whether a field is indexed or not, so whether that field is searchable. Stored determines if the field is stored in the index, so whether or not it can be returned in a query. These configurations are made at the field level though across the entire index, not at the individual document level. So perhaps a document should just contain `Fields` and the determination of whether it will be indexed and/or stored is made in the `IndexWriter`?

Reply:
Re: the distinction of stored & indexed fields at Doc v Index - effectively comes down to if we want to make the Index schema aware or not. If we treat all fields in the document as the same, we'd have to make the Index aware of which ones we'd need to index. Our underlying problem is people can add whatever tag name they'd like for a metric, so by default all fields would have to be indexed, and some  'known' ones per Index would be overriden to be only stored. It feels cleaner to have the distinction at the document level instead of some scheme like that. 

Re analyzed fields: Analysers/Tokenisers add a lot of complexity. We can come back to this if ever need them.

Reply:
In regards to supporting a schema or not, I actually think it would be preferable to have defaults for the fields with overrides setup at initialization for some. If a field is indexed for one document but not for others I think that can lead to confusing results.

Not suggesting we add Analysis at this stage :)

Reply:
Either way the burden to specify what to index lies on the user. I'd prefer going the schema aware route too but I'm concerned about indexing redundant fields. 

Considering that metric tag names can being anything, we have to index all fields by default. Say everything's working fine and now we want to tack on a stored field to be returned with a metric  (e.g. an extra R2 policy). People are just going to add that field to the document and move on. Adds a lot of work for them to go make the change to the index to say all fields with this name are stored fields, etc. It's much easier to encourage the correct behaviour by allowing them to specify index v stored at the document level. 

TBH, I'd change my tune if we could change the default to be un-indexed. At that point, it'd be much better for the system to be schema aware at the Index level.

Reply:
I think Elasticsearch handles pretty cleanly already. When an index is created you specify a template which can include fields and their attributes and also defaults. Then when you index a document, if it has a field in the schema, the attributes for that schema are applied. Otherwise a new field is created in the schema and the defaults are applied. New fields can added to the schema after the fact as long as they haven't been added to the schema already.

Reply:
I don't think this something we need to decide right now. Lets start by supporting only indexed fields, i.e. no stored fields. I'll change `Document` to the following definition:

```
type Document interface {
  ID ID 
  Fields []Field
}
```

We can come back to this discussion when we want to support stored fields.

Reply:
Sounds good to me. Just to be clear though, I think whether a field is stored and whether it is indexed are orthogonal concepts. A field can be
- (stored, indexed) in which case it can be searched for and returned in results
- (not stored, indexed) in which case it can be searched for but not returned in results, I see this as being something like the contents of the book, you might store metadata about the book and a path to where it can be found, but there's probably no need to include the contents of the book in an index itself if it can be found elsewhere when needed
- (stored, not indexed) this might be metadata that would never be searched for but is helpful to see in results
- (not stored, not indexed) this is just weird

Anyway, just wanted to clarify my own thoughts. Starting with every field being (stored, indexed) sounds good to me.